### PR TITLE
[REACTOS] Add support for compiling with GCC 13

### PR DESCRIPTION
**Pull request by tkreuzer**
## Purpose

This PR makes it possible to compile ros with GCC 13. It will still create warnings, so for x86 it requires to remove `-Werror`

## Proposed changes
- Add some compiler flags for GCC 13
- Add imported library winpthread
- Add support library for GCC's libsupc++ and libstdc++

### CODE REVIEW

File: sdk/cmake/gcc.cmake [hunk 1]
```diff
@@ -552,12 +552,24 @@ else()
...
+    add_library(libwinpthread INTERFACE)
+endif()
+
 add_library(libgcc STATIC IMPORTED)
 execute_process(COMMAND ${GXX_EXECUTABLE} -print-file-name=libgcc.a OUTPUT_VARIABLE LIBGCC_LOCATION)
 string(STRIP ${LIBGCC_LOCATION} LIBGCC_LOCATION)
 set_target_properties(libgcc PROPERTIES IMPORTED_LOCATION ${LIBGCC_LOCATION})
 # libgcc needs kernel32 imports, a CRT and msvcrtex
-target_link_libraries(libgcc INTERFACE libkernel32 libmsvcrt msvcrtex)
+target_link_libraries(libgcc INTERFACE libwinpthread libkernel32 libmsvcrt msvcrtex)
```
Comment (HBelusca) [inline sdk/cmake/gcc.cmake:572]: geez GCC... why the hell do they need winpthread?

File: sdk/lib/gcc-compat/__throw_out_of_range_fmt.cpp [hunk 2]
```diff
@@ -0,0 +1,42 @@
+/*
+ * PROJECT:     GCC c++ support library
+ * LICENSE:     MIT (https://spdx.org/licenses/MIT)
+ * PURPOSE:     __throw_out_of_range_fmt implementation
+ * COPYRIGHT:   Copyright 202024 Timo Kreuzer <timo.kreuzer@reactos.org>
```
Comment (HBelusca) [inline sdk/lib/gcc-compat/__throw_out_of_range_fmt.cpp:5]: Wow it's far ahead in the future! (And same observation for the other files as well)

File: sdk/lib/gcc-compat/StringCchCopyNA.c [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+#include <windows.h>
+#define StringCchCopyNA _StringCchCopyNA
+#include <strsafe.h>
+
+#undef StringCchCopyNA
+
+HRESULT
+WINAPI
+StringCchCopyNA(
```
Comment (HBelusca) [inline sdk/lib/gcc-compat/StringCchCopyNA.c:16]: Which code in there requires this function? It doesn't appear to be used, at least explicitly.

File: sdk/cmake/gcc.cmake (hunk 1 above)
Comment (tkreuzer) [reply sdk/cmake/gcc.cmake:572]: For threading support in their runtime library I guess. The better alternative on Windows would be MCF Gthread, but that requires Windows 7.

File: sdk/lib/gcc-compat/StringCchCopyNA.c (hunk 3 above)
Comment (tkreuzer) [reply sdk/lib/gcc-compat/StringCchCopyNA.c:16]: GCC's c++ library.

File: sdk/cmake/gcc.cmake (hunk 1 above)
Comment (learn-more) [reply sdk/cmake/gcc.cmake:572]: I do hope all this gcc nastiness is static, and our binaries do not actually depend on some shit like winpthread.dll or so?

**Review by learn-more — APPROVED** [review]

File: sdk/cmake/gcc.cmake (hunk 1 above)
Comment (HBelusca) [reply sdk/cmake/gcc.cmake:572]: That's what I was wondering too.

Comment (simonelombardo) [reply sdk/cmake/gcc.cmake:572]: Luckily, libwinpthread and libgcc_dw2 / libgcc_seh2 can be statically linked to the produced executable (it's actually a compiler option of the configure step of gcc buildscript that ends up generating libwinpthread.dll.a or libwinpthread.a).

File: sdk/lib/gcc-compat/StringCchCopyNA.c (hunk 3 above)
Comment (HBelusca) [reply sdk/lib/gcc-compat/StringCchCopyNA.c:16]: strange that they use it (I would have expected they use the posix string funcs instead :rofl: )

**Review by HBelusca — APPROVED** [review]
OK but Timo please fix that `Copyright 202024 ` date in the three new files.

File: sdk/cmake/gcc.cmake [hunk 4]
```diff
@@ -54,6 +54,24 @@ if(CMAKE_C_COMPILER_ID STREQUAL "GNU")
...
+        add_compile_options(-fno-builtin-sin)
+        add_compile_options(-fno-builtin-cos)
+        add_compile_options(-fno-builtin-sincos)
+        add_compile_options(-fno-builtin-pow)
+        add_compile_options(-fno-builtin-sqrt)
+        add_compile_options(-fno-builtin-sqrtf)
+        add_compile_options(-fno-builtin-floor)
+        add_compile_options(-fno-builtin-floorf)
+        add_compile_options(-fno-builtin-ceilf)
+        add_compile_options(-fno-builtin-ceil)
```
Comment (SergeGautherie) [inline sdk/cmake/gcc.cmake:68]: Sort them. Want to merge them into "1" line?

**Review by cbialorucki — APPROVED** [review]
Just fix the copyright dates

**Comment by wjk** [comment]
If I understand correctly, this new library is required for any code compiled with GCC to run. winpthread dependency aside, does this not mean that everything now depends on kernel32.dll and msvcrt.dll, including kernel components and DLLs lower than kernel32 and msvcrt? Won’t this cause problems?

File: sdk/cmake/gcc.cmake [hunk 5]
```diff
@@ -534,19 +552,31 @@ else()
...
-target_link_libraries(libgcc INTERFACE libkernel32 libmsvcrt msvcrtex)
+target_link_libraries(libgcc INTERFACE libwinpthread libkernel32 libmsvcrt msvcrtex)
 
 add_library(libsupc++ STATIC IMPORTED GLOBAL)
 execute_process(COMMAND ${GXX_EXECUTABLE} -print-file-name=libsupc++.a OUTPUT_VARIABLE LIBSUPCXX_LOCATION)
 string(STRIP ${LIBSUPCXX_LOCATION} LIBSUPCXX_LOCATION)
 set_target_properties(libsupc++ PROPERTIES IMPORTED_LOCATION ${LIBSUPCXX_LOCATION})
 # libsupc++ requires libgcc
-target_link_libraries(libsupc++ INTERFACE libgcc)
+target_link_libraries(libsupc++ INTERFACE libgcc stdc++compat)
```
Comment (SergeGautherie) [inline sdk/cmake/gcc.cmake:579]: Update comment too?

File: sdk/cmake/gcc.cmake [hunk 6]
```diff
@@ -534,19 +552,31 @@ else()
...
+    add_library(libwinpthread INTERFACE)
+endif()
+
 add_library(libgcc STATIC IMPORTED)
 execute_process(COMMAND ${GXX_EXECUTABLE} -print-file-name=libgcc.a OUTPUT_VARIABLE LIBGCC_LOCATION)
 string(STRIP ${LIBGCC_LOCATION} LIBGCC_LOCATION)
 set_target_properties(libgcc PROPERTIES IMPORTED_LOCATION ${LIBGCC_LOCATION})
 # libgcc needs kernel32 imports, a CRT and msvcrtex
-target_link_libraries(libgcc INTERFACE libkernel32 libmsvcrt msvcrtex)
+target_link_libraries(libgcc INTERFACE libwinpthread libkernel32 libmsvcrt msvcrtex)
```
Comment (SergeGautherie) [inline sdk/cmake/gcc.cmake:572]: Update comment too?

**Review by DarkFire01 — APPROVED** [review]

**Comment by HBelusca** [comment]
I wonder whether it is possible to add a temporary github action that makes use of GCC13, to see how things go.

File: sdk/lib/gcc-compat/StringCchCopyNA.c [hunk 7]
```diff
@@ -0,0 +1,23 @@
+/*
+ * PROJECT:     GCC c++ support library
+ * LICENSE:     MIT (https://spdx.org/licenses/MIT)
+ * PURPOSE:     StringCchCopyNA implementation
+ * COPYRIGHT:   Copyright 202024 Timo Kreuzer <timo.kreuzer@reactos.org>
```
Comment (oleg-dubinskiy) [inline sdk/lib/gcc-compat/StringCchCopyNA.c:5]: Fix the copyright date here as follows:
`COPYRIGHT:   Copyright 2024 Timo Kreuzer <timo.kreuzer@reactos.org>`

File: sdk/lib/gcc-compat/__mingw_vsprintf.c [hunk 8]
```diff
@@ -0,0 +1,13 @@
+/*
+ * PROJECT:     GCC c++ support library
+ * LICENSE:     MIT (https://spdx.org/licenses/MIT)
+ * PURPOSE:     __mingw_vsprintf implementation
+ * COPYRIGHT:   Copyright 202024 Timo Kreuzer <timo.kreuzer@reactos.org>
```
Comment (oleg-dubinskiy) [inline sdk/lib/gcc-compat/__mingw_vsprintf.c:5]: Same here...

File: sdk/lib/gcc-compat/__throw_out_of_range_fmt.cpp (hunk 2 above)
Comment (oleg-dubinskiy) [inline sdk/lib/gcc-compat/__throw_out_of_range_fmt.cpp:5]: ...and here.

File: sdk/cmake/gcc.cmake [hunk 9]
```diff
@@ -534,19 +552,31 @@ else()
...
 
 add_library(libsupc++ STATIC IMPORTED GLOBAL)
 execute_process(COMMAND ${GXX_EXECUTABLE} -print-file-name=libsupc++.a OUTPUT_VARIABLE LIBSUPCXX_LOCATION)
 string(STRIP ${LIBSUPCXX_LOCATION} LIBSUPCXX_LOCATION)
 set_target_properties(libsupc++ PROPERTIES IMPORTED_LOCATION ${LIBSUPCXX_LOCATION})
 # libsupc++ requires libgcc
```
Comment (oleg-dubinskiy) [inline sdk/cmake/gcc.cmake:578]: And update this comment as follows:
`# libsupc++ requires libgcc and stdc++compat`

File: sdk/cmake/gcc.cmake [hunk 10]
```diff
@@ -534,19 +552,31 @@ else()
...
+    target_link_libraries(libwinpthread INTERFACE libkernel32 libmsvcrt msvcrtex)
+else()
+    add_library(libwinpthread INTERFACE)
+endif()
+
 add_library(libgcc STATIC IMPORTED)
 execute_process(COMMAND ${GXX_EXECUTABLE} -print-file-name=libgcc.a OUTPUT_VARIABLE LIBGCC_LOCATION)
 string(STRIP ${LIBGCC_LOCATION} LIBGCC_LOCATION)
 set_target_properties(libgcc PROPERTIES IMPORTED_LOCATION ${LIBGCC_LOCATION})
 # libgcc needs kernel32 imports, a CRT and msvcrtex
```
Comment (oleg-dubinskiy) [inline sdk/cmake/gcc.cmake:571]: ... and this one as follows:
`# libgcc needs winpthread, kernel32, a CRT and msvcrtex imports`

File: sdk/cmake/gcc.cmake (hunk 4 above)
Comment (oleg-dubinskiy) [inline sdk/cmake/gcc.cmake:68]: Also, sort these in alphabetical order.

**Review by oleg-dubinskiy — APPROVED** [review]
Some nit-picks, LGTM otherwise.

File: sdk/cmake/gcc.cmake [hunk 11]
```diff
@@ -54,6 +54,24 @@ if(CMAKE_C_COMPILER_ID STREQUAL "GNU")
...
+        add_compile_options(-fno-builtin-cos)
+        add_compile_options(-fno-builtin-floor)
+        add_compile_options(-fno-builtin-floorf)
+        add_compile_options(-fno-builtin-pow)
+        add_compile_options(-fno-builtin-sin)
+        add_compile_options(-fno-builtin-sincos)
+        add_compile_options(-fno-builtin-sqrt)
+        add_compile_options(-fno-builtin-sqrtf)
+    endif()
+    if(CMAKE_C_COMPILER_VERSION VERSION_GREATER_EQUAL 13)
```
Comment (SergeGautherie) [inline sdk/cmake/gcc.cmake:70]: Shouldn't the other changes (winpthread, gcc-compat, ...) too depend on GCC version? As in unneeded for RosBE's GCC 8.4.0.

**Review by learn-more — APPROVED** [review]

**Pull request merged**